network/gossip.go

```diff
@@ -53,32 +58,62 @@ func (t *Topic) Subscribe(handler func(obj interface{})) error {
 	return nil
 }
 
+func (t *Topic) Close() error {
+	close(t.unsubscribeCh)
+	t.wg.Wait()
+
+	return t.topic.Close()
+}
+
 func (t *Topic) readLoop(sub *pubsub.Subscription, handler func(obj interface{})) {
 	ctx, cancelFn := context.WithCancel(context.Background())
 
+	var unsubscribe int32 = 0
+
+	workqueue := make(chan proto.Message, runtime.NumCPU())
+
+	t.wg.Add(1)
+
 	go func() {
-		<-t.closeCh
+		<-t.unsubscribeCh
+		atomic.StoreInt32(&unsubscribe, 1)
+
+		close(workqueue)
+
 		cancelFn()
+		sub.Cancel()
+		t.wg.Done()
 	}()
 
-	for {
+	for i := 0; i < runtime.NumCPU(); i++ {
```

Comment:
Would it be any race condition here?

Reply:
general worker pool size will use this, maybe need clamp, too many goroutines can increase CPU contention